### ISSUE TITLE
Clarify needed permissions for containers

### ISSUE THREAD

**Issue by Jegelewicz** [ArctosDB/arctos#7127]
# Latest 

Access proposal here: https://github.com/ArctosDB/arctos/issues/7127#issuecomment-1969221308

-----------

# The Spreadsheet

Don't worry about current anything, tell me what you want at https://docs.google.com/spreadsheets/d/1Dha9tra2W_h8Rn9tJN6dSMXvy0Nl0Lc8dCLDt2FEN10/edit?usp=sharing

-----------------

# Original

[Bulk edit container](https://arctos.database.museum/loaders/bulkEditContainer.cfm) asserts the following:

>  Required to Insert: manage_container
> 
>     Many loaders are accessible via "Data Entry Extras," bot users, or applications other than the management tool. Users who can insert can generally view or download their own records, but may not make further updates.
> 
> Required to Update: manage_records,manage_container
> 
>     Users with manage roles can generally review, load, download, or delete records by users with whom they share collections

However, a user (rpianka) with access to the appropriate collection and the required roles

![image](https://github.com/ArctosDB/arctos/assets/5725767/c2b19c06-f834-4aa8-a4dd-d3ae9aef4505)

is not able to load data to the tool.

Can anyone tell me the roles required and then can we update the tool documentation to provide the correct roles required to use this tool?

**Comment by dustymc** [comment]
I don't see any obvious problems and need more information. (Maybe an update hadn't cleared the cache?)

**Comment by Jegelewicz** [comment]
See https://github.com/ArctosDB/data-migration/issues/1517#issuecomment-1856148408

**Comment by falco-rk** [comment]
@dustymc @Jegelewicz I am having the same issue again with a new user (miabender). I want her to be able to edit containers and based on the documentation, she should be able to. 

We have logged in/out many times, cleared the cache in both firefox and chrome and she still gets "insufficient access" when she goes to autoload her data in the container bulkloader. Please help.

![Screenshot 2024-01-19 at 12 36 33 PM](https://github.com/ArctosDB/arctos/assets/49788908/ca09651e-8208-40d8-8740-49fc25de22a5)

**Comment by dustymc** [comment]
>container bulkloader

That's create (should be a rare and scary thing), not edit, and it needs admin_container. The form will tell you what's necessary:

 Required to Update: admin_container

**Comment by falco-rk** [comment]
I meant container bulk edit, not container bulkloader

**Comment by dustymc** [comment]
bulk edit also requires admin_container, but I'm not sure if it should or not. At some level edit access is edit access, or maybe the ability to ~~mangle~~ edit a million containers with a click needs some special protection. Help?!

**Comment by campmlc** [comment]
I agree we need to revist this, as it is not clear from the documentation, and some of the role restrictions are clearly problematic for collections.

**Comment by dustymc** [comment]
https://docs.google.com/spreadsheets/d/1Dha9tra2W_h8Rn9tJN6dSMXvy0Nl0Lc8dCLDt2FEN10/edit?usp=sharing

spreadsheet to document access

**Comment by Jegelewicz** [comment]
From today's discussion in the Object Tracking meeting

role | actions | DB roles 
-- | -- | -- 
view | see stuff - edit nothing | read 
move | update only parent_container_id | update
manage | create, edit, move, view | create
admin | series, barcode swaps | delete

@dustymc is this something we can work with for object tracking roles? What is needed from us to sort this out and set up appropriate roles?

**Comment by dustymc** [comment]
Yes, that's actionable, and if I move a bit of CF code to a procedure I think I can do it all in the DB (where it can scale).

Suggest dropping the DB roles column, I think that's just a distraction to most.

I think this needs a new Issue (or just a title change and email? I'm happy with whatever @Jegelewicz wants here). We're not longer just clarifying. I'm most concerned with the social aspects of this - and only that it'll be misunderstood, not that this isn't a logical approach. I think this needs as much exposure as possible.

I would hope that part of the plan would be to use these roles (plus manage_records to access bulk tools, perhaps) in the forms themselves, so I think this does accomplish the clarification goals, and makes it easy for me to stay with that. (Some of the forms will need rebuilt to follow, but they've needed a rebuild for a while anyway.)

I like it, if everyone else can/will live with it.

One precursor issue might be the use of institution_acronym - someone (@campmlc ?) mentioned that this isn't great, that's understandable, I don't care and can change it (presumably to guid_prefix), but the RLS/VPD policy is written around it; changing it would change the below (probably most users don't have access to all collections which share an institution), and that would change the code I need to write. If we are going to change that, it would be very good to do it before this if there's any way to do that. 

Actually, considering the social aspects - that would require two-way communication with the collections - let's update "very good" to "necessary." If this needs to change, someone file the Issue and escort it through, if that doesn't happen then we have to consider this dealt with. OK?

Here's a slight revision of your table. 

 
role | actions | dependencies
-- | -- | --
view_container | see everything, change nothing | none
move_container | update only parent_container_id | view_container
manage_container | create, edit, update, delete anything | move_container
admin_container | manage container series data, execute (to be written) function barcode_swapper | manage_container


* "Everything" is always constrained by the aforementioned controls, no user can access anything in institutions to which they don't have access.
* "Anything" is constrained by table rules - eg, 'update anything' doesn't allow changing barcode (those are ALWAYS meant to be scans), must stay within weight constraints, etc.
* I don't necessarily have to structure this as hierarchical/cascading/dependent roles - more_container could have view_container built in - but doing so would simplify/modularize things, and I can't see any realistic scenario in which these don't accumulate with experience.
* There are no "secondary guardrails" built into this. A user can create containers or they can't; that includes creating buildings (probably very rare), setting up position-holding containers (things that require significant understanding and are difficult and tedious to un-do), or creating the positions within those containers (near impossible to get wrong if the parent is set up correctly and there's any HINT of a procedure). I think I like that, even though it's not quite where we're coming from: the collections just have to decide if some user can be trusted to operate within their training or not.

And here's the full table definition, in case that's handy:

```
arctosprod@arctos>> \d container
                                    Table "core.container"
            Column             |            Type             | Collation | Nullable | Default 
-------------------------------+-----------------------------+-----------+----------+---------
 container_id                  | integer                     |           | not null | 
 parent_container_id           | integer                     |           | not null | 
 container_type                | character varying(30)       |           | not null | 
 label                         | character varying(255)      |           | not null | 
 description                   | character varying(255)      |           |          | 
 last_date                     | timestamp without time zone |           |          | 
 container_remarks             | character varying(4000)     |           |          | 
 barcode                       | character varying(50)       |           |          | 
 print_fg                      | smallint                    |           |          | 
 width                         | double precision            |           |          | 
 height                        | double precision            |           |          | 
 length                        | double precision            |           |          | 
 institution_acronym           | character varying(20)       |           | not null | 
 number_rows                   | integer                     |           |          | 
 number_columns                | integer                     |           |          | 
 orientation                   | character varying(25)       |           |          | 
 positions_hold_container_type | character varying(25)       |           |          | 
 last_update_tool              | character varying(255)      |           |          | 
 weight                        | numeric                     |           |          | 
 weight_units                  | character varying           |           |          | 
 weight_capacity               | numeric                     |           |          | 
 weight_capacity_units         | character varying           |           |          | 
Indexes:
    "container_pkey" PRIMARY KEY, btree (container_id)
    "iu_container_barcode_inst" UNIQUE, btree (barcode, institution_acronym)
    "ix_container_barcode" btree (barcode)
    "ix_container_label" btree (label)
    "ix_container_pcid" btree (parent_container_id)
    "ix_container_printfg" btree (print_fg)
    "ix_upr_container_barcode" btree (upper(barcode::text))
Check constraints:
    "ck_barcode_noprint" CHECK (checkfreetext(barcode))
    "ck_container_remarks_noprint" CHECK (checkfreetext(container_remarks))
    "ck_description_noprint" CHECK (checkfreetext(description))
    "ck_height" CHECK (COALESCE(height, 1::double precision) > 0::double precision)
    "ck_label_noprint" CHECK (checkfreetext(label))
    "ck_length" CHECK (COALESCE(length, 1::double precision) > 0::double precision)
    "ck_weight" CHECK (weight > 0::numeric)
    "ck_weight_capacity" CHECK (weight_capacity > 0::numeric)
    "ck_width" CHECK (COALESCE(width, 1::double precision) > 0::double precision)
Foreign-key constraints:
    "container_weight_capacity_units_fkey" FOREIGN KEY (weight_capacity_units) REFERENCES ctweight_units(weight_units)
    "container_weight_units_fkey" FOREIGN KEY (weight_units) REFERENCES ctweight_units(weight_units)
    "fk_ctcontainer_type" FOREIGN KEY (container_type) REFERENCES ctcontainer_type(container_type)
    "fk_institution_acronym" FOREIGN KEY (institution_acronym) REFERENCES institution(institution_acronym)
    "fk_psn_hld_ctr_typ" FOREIGN KEY (positions_hold_container_type) REFERENCES ctcontainer_type(container_type)
Referenced by:
    TABLE "container_environment" CONSTRAINT "fk_cenv_container_id" FOREIGN KEY (container_id) REFERENCES container(container_id)
    TABLE "coll_obj_cont_hist" CONSTRAINT "fk_collobjconthist_container" FOREIGN KEY (container_id) REFERENCES container(container_id)
    TABLE "container_check" CONSTRAINT "fk_containercheck_container" FOREIGN KEY (container_id) REFERENCES container(container_id)
    TABLE "shipment" CONSTRAINT "fk_shipment_container" FOREIGN KEY (container_id) REFERENCES container(container_id)
    TABLE "trans_container" CONSTRAINT "fk_transcontainer_container" FOREIGN KEY (container_id) REFERENCES container(container_id)
Policies:
    POLICY "rls_container_policy"
      USING (((institution_acronym)::text IN ( SELECT unnest(get_my_institutions()) AS unnest)))
Triggers:
    get_container_history AFTER INSERT OR DELETE OR UPDATE ON container FOR EACH ROW EXECUTE FUNCTION trigger_fct_get_container_history()
    trg_cont_defdate BEFORE INSERT OR UPDATE ON container FOR EACH ROW EXECUTE FUNCTION trigger_fct_trg_cont_defdate()
    trg_container_bupins BEFORE INSERT OR UPDATE ON container FOR EACH ROW EXECUTE FUNCTION trigger_fct_trg_container_bupins()
    trg_container_content_control BEFORE INSERT OR UPDATE ON container FOR EACH ROW EXECUTE FUNCTION trigger_fct_trg_container_content_control()
    trg_container_weight_control AFTER INSERT OR UPDATE ON container FOR EACH ROW EXECUTE FUNCTION trigger_fct_trg_container_weight_control()
```

**Comment by campmlc** [comment]
Can you clarify what you mean by switching from institution acronym to guid prefix?

**Comment by dustymc** [comment]
>Can you clarify what you mean by switching from institution acronym to guid prefix?

The table is currently....

```
arctosprod@arctos>> \d container
                                    Table "core.container"
            Column             |            Type             | Collation | Nullable | Default 
-------------------------------+-----------------------------+-----------+----------+---------
 container_id                  | integer                     |           | not null | 
...
 institution_acronym           | character varying(20)       |           | not null | 
...
Indexes:
...
    "iu_container_barcode_inst" UNIQUE, btree (barcode, institution_acronym)
...
Foreign-key constraints:
    ...
    "fk_institution_acronym" FOREIGN KEY (institution_acronym) REFERENCES institution(institution_acronym)
...
Policies:
    POLICY "rls_container_policy"
      USING (((institution_acronym)::text IN ( SELECT unnest(get_my_institutions()) AS unnest)))
...
```

Those could all be changed to involve collection.collection_id (that is, collections - or some proxy to them - rather than institutions, which don't actually exist except as semi-uncontrolled text strings within collections).

Making such a change would require an Issue, and a commitment by someone to see it through.

**Comment by campmlc** [comment]
So does that mean that container permissions would be assigned at the guid prefix collection level vs institution level? And the commitment would involve making sure collection managers understand the responsibilities involved? Or am I misunderstanding?

**Comment by dustymc** [comment]
>container permissions would be assigned at the guid prefix collection level vs institution level?

Yes - they'd follow the VPD model of everything else (eg catalog records) instead of needing their own institution-based approach.

>commitment would involve making sure collection managers understand

That's the start, but the real issue will be updating containers from being associated with institution to collections. I can't make that decision by myself, and I can't start this until every container is mapped to a collection.

**Comment by dustymc** [comment]
>Possible to change

Yes, delete can be moved, I'll update https://github.com/ArctosDB/arctos/issues/7127#issuecomment-1969221308

**Comment by dustymc** [comment]
The potential precursor has an Issue, https://github.com/ArctosDB/arctos/issues/7466, which is now blocking this. I'll update labels and I'm going to remove some things which should be discussed there from this so we can stay focused.

**Comment by falco-rk** [comment]
So I'm having the same issue again where a new user, SyringaBarenti, can't update containers. 

She has these permissions: admin_container, coldfusion_user, create_records, data_entry, manage_container, manage_records, manage_specimens, manage_transactions, syringabarenti, view_container

I can do it and I have these permissions: admin_container, coldfusion_user, create_records, data_entry, manage_agents, manage_collection, manage_container, manage_locality, manage_publications, manage_records, manage_specimens, manage_taxonomy, manage_transactions, rosiepoulin, view_container

@happiah-madson @Syringab @AnnaEaton

**Comment by dustymc** [comment]
I'm not seeing anything very obvious in the logs, please elaborate - what is she trying and what's happening?

**Comment by falco-rk** [comment]
Never mind, it worked this time. I think she tried to do it before Arctos had updated her permissions.

**Comment by campmlc** [comment]
From https://github.com/ArctosDB/dev/issues/93#issuecomment-2452485552
There is a problem with the update_container definition. This tool was requested as a means of allowing beginning and intermediate students to scan barcodes into container positions, with the ability to only "update" the position, without having to provide manage specimens or manage container. Yet the role definition says: "Intermediate user; primarily intended to allow scanning containers into containers, but allows UPDATE on all fields in table container (excepting barcode). Prerequisite: manage_container."
The prerequisite should be changed to "view_container", not "manage_container".

Same problem with manage_container definition and role: Create, delete and edit containers. Prerequisite: view_container
This should be changed to prerequisite = update_container.

**Comment by Jegelewicz** [comment]
From today's Genomic Resources Meeting:

https://arctos.database.museum/tools/bulkCreateContainerPositions.cfm

either should be added to manage_container OR have it's own permission level

**Comment by campmlc** [comment]
Following up with this: 
Today I had to give a student admin container permissions so she could use Bulk Edit Container. Can we please move this and create container tools out of admin role as soon as possible, as discussed last week in the Genomic Resources meeting? Putting in manage container pending a new role if necessary is fine, but I need to be able to have students use these tools without admin permissions right away. This is holding up work, and it terrifies me to have to give out admin roles. @mkoo @dustymc

**Comment by dustymc** [comment]
>admin container permissions so she could use Bulk Edit Container

I don't see the connection, elaborate? (But maybe not here, I'm closing this wandering issue, please open specific requests if there's something not addressed by a dev issue or long-since completed.)
